pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -273,21 +263,27 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListeners(env model.Env
 				useRemoteAddress: false,
 				direction:        http_conn.INGRESS,
 			}
-			if listenerOpts.tlsMultiplexed {
-				listenerOpts.filterChainOpts = []*filterChainOpts{
-					{
-						httpOpts:          httpOpts,
-						transportProtocol: authn.EnvoyRawBufferMatch,
-					}, {
-						httpOpts:          httpOpts,
-						transportProtocol: authn.EnvoyTLSMatch,
-					},
+			found := false
+			for _, p := range configgen.Plugins {
+				if authnPolicy, ok := p.(authn.Plugin); ok {
+					matches, tls, requireTLSInspector := authnPolicy.SetupFilterChains(
+						env.Mesh, env.IstioConfigStore, instance.Service.Hostname, instance.Endpoint.ServicePort)
+					listenerOpts.requireTLSInspector = requireTLSInspector
+					for i, match := range matches {
+						listenerOpts.filterChainOpts = append(listenerOpts.filterChainOpts, &filterChainOpts{
+							httpOpts:   httpOpts,
+							match:      match,
+							tlsContext: tls[i],
+						})
+					}
+					found = true
```

Comment:
why can't you wrap this entire code in a callback? like OnFilterChains() callback, thats implemented only by Auth? Pass in the listener opts, etc.. and the plugin can return the filterchain opts.

Reply:
Sounds like adding a new method to Plugin interface and only implemented in authn plugin for now.

I'll do do that if that's what you refer to.